// grammy-inline-menu/dist/source/keyboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Keyboard = void 0;
const node_buffer_1 = require("node:buffer");
const path_js_1 = require("./path.js");
const generic_types_js_1 = require("./generic-types.js");
function isRow(entry) {
    return Array.isArray(entry);
}
function isCallbackButtonTemplate(kindOfButton) {
    return 'text' in kindOfButton && 'relativePath' in kindOfButton;
}
class Keyboard {
    constructor() {
        Object.defineProperty(this, "_entries", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
    }
    addCreator(creator) {
        this._entries.push(creator);
    }
    add(joinLastRow, ...buttons) {
        const lastEntry = this._entries.slice(-1)[0];
        if (joinLastRow && isRow(lastEntry)) {
            lastEntry.push(...buttons);
            return;
        }
        this._entries.push([...buttons]);
    }
    async render(context, path) {
        const arrayOfRowArrays = await Promise.all(
        // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
        this._entries.map(async (o) => entryToRows(o, context, path)));
        const rows = arrayOfRowArrays
            .flat(1)
            .map(row => renderRow(row, path))
            .filter(o => o.length > 0);
        return rows;
    }
}
exports.Keyboard = Keyboard;
async function entryToRows(entry, context, path) {
    if (typeof entry === 'function') {
        return entry(context, path);
    }
    const buttonsInRow = await Promise.all(entry.map(async (button) => typeof button === 'function' ? button(context, path) : button));
    const filtered = buttonsInRow.filter((0, generic_types_js_1.filterNonNullable)());
    return [filtered];
}
function renderRow(templates, path) {
    return templates
        .map(template => isCallbackButtonTemplate(template) ? renderCallbackButtonTemplate(template, path) : template);
}
function renderCallbackButtonTemplate(template, path) {
    const absolutePath = (0, path_js_1.combinePath)(path, template.relativePath);
    const absolutePathLength = node_buffer_1.Buffer.byteLength(absolutePath, 'utf8');
    if (absolutePathLength > 64) {
        throw new Error(`callback_data only supports 1-64 bytes. With this button (${template.relativePath}) it would get too long (${absolutePathLength}). Full path: ${absolutePath}`);
    }
    return {
        text: template.text,
        callback_data: absolutePath,
    };
}
