// grammy-inline-menu/dist/source/prefix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prefixText = exports.prefixEmoji = exports.emojiFalse = exports.emojiTrue = void 0;
exports.emojiTrue = '✅';
exports.emojiFalse = '🚫';
/**
 * Prefixes the text with a true / false emoji.
 * Can also be used with custom prefixes.
 * @param text text which should receive the prefix.
 * @param prefix true / false or a custom (string) prefix.
 * @param options optional options to customize emojis
 */
function prefixEmoji(text, prefix, options = {}) {
    const internalOptions = {
        ...options,
        prefixTrue: options.prefixTrue ?? exports.emojiTrue,
        prefixFalse: options.prefixFalse ?? exports.emojiFalse,
    };
    const prefixContent = applyOptionsToPrefix(prefix, internalOptions);
    return prefixText(text, prefixContent);
}
exports.prefixEmoji = prefixEmoji;
function applyOptionsToPrefix(prefix, options) {
    const { prefixFalse, prefixTrue, hideFalseEmoji, hideTrueEmoji, } = options;
    if (prefix === true) {
        if (hideTrueEmoji) {
            return undefined;
        }
        return prefixTrue;
    }
    if (prefix === false) {
        if (hideFalseEmoji) {
            return undefined;
        }
        return prefixFalse;
    }
    return prefix;
}
/**
 * Prefixes the text with the prefix.
 * If the prefix is undefined or '' the raw text is returned.
 * @param text text which should receive the prefix.
 * @param prefix prefix to end up in front of the text.
 */
function prefixText(text, prefix) {
    if (!prefix) {
        return text;
    }
    return `${prefix} ${text}`;
}
exports.prefixText = prefixText;
